### PR TITLE
TL/UCP: remove memcpy in last SRA step

### PR DESCRIPTION
**Pull request by Sergei-Lebedev**
## What
Avoid memory copy in last step of reduce scatter.

## How ?
In last step of reduce scatter local reduce operation is split into strided reduce of received data and reduce of result of that operation and local data.

### CODE REVIEW

File: src/utils/ucc_dt_reduce.h [hunk 1]
```diff
@@ -33,7 +33,7 @@ ucc_dt_reduce_strided(void *src1, void *src2, void *dst, size_t n_vectors,
 {
     ucc_ee_executor_task_args_t eargs;
 
-    if (count == 0) {
+    if (count == 0 || n_vectors == 0) {
```
Comment (shimmybalsam) [inline src/utils/ucc_dt_reduce.h:36]: why?

File: src/components/tl/ucp/reduce_scatter/reduce_scatter_knomial.c [hunk 2]
```diff
@@ -180,25 +178,42 @@ void ucc_tl_ucp_reduce_scatter_knomial_progress(ucc_coll_task_t *coll_task)
...
+
         ucc_sra_kn_get_offset_and_seglen(count, dt_size, rank, size, radix,
                                          &offset, &local_seg_count);
-        eargs.task_type = UCC_EE_EXECUTOR_TASK_COPY;
-        eargs.copy.dst  = PTR_OFFSET(args->dst.info.buffer, offset);
-        eargs.copy.src  = task->reduce_scatter_kn.scratch;
-        eargs.copy.len  = local_seg_count * dt_size;
-        status = ucc_ee_executor_task_post(task->reduce_scatter_kn.executor, &eargs,
-                                           &task->reduce_scatter_kn.etask);
+        status = ucc_dt_reduce(local_data, rbuf,
```
Comment (vspetrov) [inline src/components/tl/ucp/reduce_scatter/reduce_scatter_knomial.c:203]: I seem to misunderstand the change. So, in case of !scratch (full pow radix case + no INPLACE) on the last reduction you do it in 2 steps: first you add all the buffers received during last step, place result in the beginning of the "rbuf" (which is dst.buffer). Then you do 1 more step in the end: reduce local data with that intermediate result and place the answer right back rbuf. Right?

How is that different from doing the entire reduce_strided with step_radix - 1?

File: src/components/tl/ucp/reduce_scatter/reduce_scatter_knomial.c [hunk 3]
```diff
@@ -180,25 +178,42 @@ void ucc_tl_ucp_reduce_scatter_knomial_progress(ucc_coll_task_t *coll_task)
...
-        ucc_knomial_pattern_next_iteration(p);
+        ucc_kn_rs_pattern_next_iter(p);
     }
 
     if (!task->reduce_scatter_kn.scratch_mc_header) {
+        ucc_knomial_pattern_prev_iteration(p);
+        get_sbuf_rbuf(p, args, task->reduce_scatter_kn.scratch, block_count,
+                      &sbuf, &rbuf);
+
+        step_radix = ucc_kn_compute_step_radix(p);
```
[the remark this answers is not in the document]
Comment (Sergei-Lebedev) [reply src/components/tl/ucp/reduce_scatter/reduce_scatter_knomial.c:189]: no, block_count is different

File: src/utils/ucc_dt_reduce.h (hunk 1 above)
Comment (Sergei-Lebedev) [reply src/utils/ucc_dt_reduce.h:36]: if n_vectors is 0 we have no data to reduce

File: src/components/tl/ucp/reduce_scatter/reduce_scatter_knomial.c [hunk 4]
```diff
@@ -71,105 +87,87 @@ void ucc_tl_ucp_reduce_scatter_knomial_progress(ucc_coll_task_t *coll_task)
...
-                                                 &offset, &local_seg_count);
-                reduce_data = PTR_OFFSET(args->dst.info.buffer, offset);
+                status = ucc_dt_reduce_strided(
+                    rbuf, PTR_OFFSET(rbuf, local_seg_count * dt_size), rbuf,
+                    step_radix - 2, local_seg_count, local_seg_count * dt_size,
+                    dt, args, 0, 0, task->reduce_scatter_kn.executor,
+                    &task->reduce_scatter_kn.etask);
+
+            } else {
+                if (task->reduce_scatter_kn.scratch_mc_header &&
```
[the remark this answers is not in the document]
Comment (Sergei-Lebedev) [reply src/components/tl/ucp/reduce_scatter/reduce_scatter_knomial.c:157]: maybe i missing something, there 3 different cases 
* last iteration and don't have scratch - partial reduction
* last iteration and have scratch  - full reduction since receive data is in scratch
* not last iteration - full reduction

File: src/components/tl/ucp/reduce_scatter/reduce_scatter_knomial.c (hunk 2 above)
Comment (Sergei-Lebedev) [reply src/components/tl/ucp/reduce_scatter/reduce_scatter_knomial.c:203]: yes, it's correct, the difference is that in this PR we do step_radix -1 reductions while in prev version it was step_radix -1 reduction plus memory copy
i see big improvement for cases where this memory copy takes a lot of time for example 
image

**Comment by vspetrov** [comment]
Discussed over phone with Sergey. He is to check if the change works correctly in the case when count % comm_size != 0.

**Comment by Sergei-Lebedev** [comment]
> Discussed over phone with Sergey. He is to check if the change works correctly in the case when count % comm_size != 0.

@vspterov, I've changed the way offsets are calculated please have a look

File: src/components/tl/ucp/reduce_scatter/reduce_scatter_knomial.c [hunk 5]
```diff
@@ -15,14 +15,57 @@
...
+            correct destination directly
+        */
+        local_seg_offset /= dt_size;
+        if (local_seg_offset <= block_count || local_seg_count == 0) {
+            *sbuf = scratch;
+            *rbuf = PTR_OFFSET(*sbuf, block_count * dt_size);
+        } else {
+            offset = (local_seg_offset - block_count) % local_seg_count;
+            *sbuf = scratch;
+            *rbuf = PTR_OFFSET(*sbuf, (block_count + offset) * dt_size);
```
Comment (vspetrov) [inline src/components/tl/ucp/reduce_scatter/reduce_scatter_knomial.c:53]: looks good to me. just one question: if scratch is actual user rbuf (no extra + no_inplace case) do we always have enough space after "block_count" to recv segement. I think so but haven't proved on paper. Maybe add assert, like:
block_ount + offset + local_seg_count <= count.

**Review by vspetrov — APPROVED** [review]

File: src/components/tl/ucp/reduce_scatter/reduce_scatter_knomial.c (hunk 5 above)
Comment (Sergei-Lebedev) [reply src/components/tl/ucp/reduce_scatter/reduce_scatter_knomial.c]: I think I checked all the corner cases and in every case where rbuf is used we have enough space. I will merge this in, let's see if any tests will fail until next release
